pkg/key/key.go

```diff
@@ -1,22 +1,109 @@
 package key
 
 import (
+	"context"
 	"fmt"
+	"io/ioutil"
+	"os"
 
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	chartv1 "github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	ClusterNameLabel        = "cluster.x-k8s.io/cluster-name"
-	ClusterWatchFilterLabel = "cluster.x-k8s.io/watch-filter"
+	ClusterNameLabel = "cluster.x-k8s.io/cluster-name"
 
 	FinalizerName = "encryption-provider-operator.finalizers.giantswarm.io"
+
+	MasterNodeLabel = "node-role.kubernetes.io/master"
 )
 
-func GetClusterIDFromLabels(t v1.ObjectMeta) string {
+func GetClusterIDFromLabels(t metav1.ObjectMeta) string {
 	return t.GetLabels()[ClusterNameLabel]
 }
 
 func SecretName(clusterName string) string {
 	return fmt.Sprintf("%s-encryption-provider-config", clusterName)
 }
+
+// GetWCK8sClient will return workload cluster k8s controller-runtime client
+func GetWCK8sClient(ctx context.Context, ctrlClient client.Client, clusterName string, clusterNamespace string) (client.Client, error) {
+	var err error
+
+	if _, err := os.Stat(tempKubeconfigFileName(clusterName)); err == nil {
+		// kubeconfig file already exists, no need to fetch and write again
+
+	} else if os.IsNotExist(err) {
+		var kubeconfig []byte
+		// kubeconfig dont exists we need to fetch it and write to file
+		var secret corev1.Secret
+		{
+			err = ctrlClient.Get(ctx, client.ObjectKey{
+				Name:      fmt.Sprintf("%s-kubeconfig", clusterName),
+				Namespace: clusterNamespace,
+			},
+				&secret)
+
+			if apierrors.IsNotFound(err) {
+				// in legacy gs the kubeconfig is in namespace equal to cluster ID
+				err = ctrlClient.Get(ctx, client.ObjectKey{
+					Name:      fmt.Sprintf("%s-kubeconfig", clusterName),
+					Namespace: clusterName,
+				},
+					&secret)
+				if err != nil {
+					return nil, err
+				}
+				kubeconfig = secret.Data["kubeConfig"]
+			} else if err != nil {
+				return nil, err
+			} else {
+				kubeconfig = secret.Data["value"]
+			}
+		}
+		err = ioutil.WriteFile(tempKubeconfigFileName(clusterName), kubeconfig, 0600)
```

Comment:
I'm not a big fan of this caching mechanism. But no big deal as long as it works

Reply:
even if I remove the caching, it kubeconfig must be written as a file in order to load it 

If you know a way how to get a controller runtime client from kubeconfig without the need of writing it to the disk I would be happy to use it